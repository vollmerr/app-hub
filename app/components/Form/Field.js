import React from 'react';
import PropTypes from 'prop-types';
import { Field as ReduxField } from 'redux-form/immutable';


function Field(WrappedComponent, requiredFunc = null) {
  return class WrappedField extends React.PureComponent {
    static propTypes = {
      required: PropTypes.bool,
      disabled: PropTypes.bool,
      validate: PropTypes.array,
      placeholder: PropTypes.string,
    };

    static defaultProps = {
      required: false,
      disabled: false,
      validate: [],
      placeholder: '',
    };

    render() {
      const {
        required,
        disabled,
        validate,
        placeholder,
        ...props
      } = this.props;

      let toValidate = [];
      let newPlaceholder;
      if (!disabled) {
        toValidate = [...validate];
        newPlaceholder = placeholder;

        if (required) {
          toValidate.push(requiredFunc);
        }
      }

      const isRequired = required && !disabled;

      const fieldProps = {
        ...props,
        disabled,
        required: isRequired,
        validate: toValidate,
        placeholder: newPlaceholder,
        component: WrappedComponent,
      };

      return (
        <ReduxField {...fieldProps} />
      );
    }
  };
}

export default Field;
