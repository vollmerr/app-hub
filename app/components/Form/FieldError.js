import styled from 'styled-components';

import theme from 'utils/theme';


const FieldError = styled.p`
  color: ${theme.redDark};
  font-size: 12px;
  margin: 0;
  padding-top: 5px;
`;

export default FieldError;
