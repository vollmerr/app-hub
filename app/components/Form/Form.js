import styled from 'styled-components';

import theme from 'utils/theme';


const Form = styled.form`
  min-height: calc(${(props) => props.vh || 100}vh - ${theme.hub.headerHeight} - 30px - ${(props) => props.margin || 0}px);
  padding: 15px;
  margin: 15px 0;
  background: ${theme.white};
`;

export default Form;
