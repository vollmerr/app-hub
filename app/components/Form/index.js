export { default as Field } from './Field';
export { default as FieldChecks } from './FieldChecks';
export { default as FieldDate } from './FieldDate';
export { default as FieldFile } from './FieldFile';
export { default as FieldRadios } from './FieldRadios';
export { default as FieldSelect } from './FieldSelect';
export { default as FieldText } from './FieldText';
export { default as FieldToggle } from './FieldToggle';
export { default as FormButtons } from './FormButtons';
export { default as FormSection } from './FormSection';
