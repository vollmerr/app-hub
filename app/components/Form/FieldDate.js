import React from 'react';
import PropTypes from 'prop-types';
import { DatePicker } from 'office-ui-fabric-react/lib/DatePicker';
import styled from 'styled-components';

import { isEmptyDate } from 'utils/validate';

import Field from './Field';

/**
 * Fix componentWillReceiveProps instead of completely rewritting...
 * - fix error displaying before selecting component / when moutned
 * - remove hard set error message / allow passing one in
 * - remove check for updating formattedDate (was false if '' value thus leaving value when 'reset')
 *
 * CHECK FOR UPDATES / IF PROBLEM FIXED:
 * https://github.com/OfficeDev/office-ui-fabric-react/blob/master/packages/office-ui-fabric-react/src/components/DatePicker/DatePicker.tsx
 */
/* istanbul ignore next */
DatePicker.prototype.componentWillReceiveProps = function (nextProps) { // eslint-disable-line
  const { formatDate, value, errorMessage } = nextProps;

  this.setState({
    errorMessage,
  });

  this.setState({
    selectedDate: value || new Date(),
    formattedDate: (formatDate && value) ? formatDate(value) : '',
  });
};

/**
 * Fix _validateTextInput instead of completely rewritting...
 * - fix error displaying when clearing values
 * - remove hard set error message / allow passing one in
 * - remove allowTextInput logic as not used
 *
 * CHECK FOR UPDATES / IF PROBLEM FIXED:
 * https://github.com/OfficeDev/office-ui-fabric-react/blob/master/packages/office-ui-fabric-react/src/components/DatePicker/DatePicker.tsx
 */
/* istanbul ignore next */
DatePicker.prototype._validateTextInput = function() { // eslint-disable-line
  const { isRequired, errorMessage } = this.props;
  const { formattedDate, isDatePickerShown } = this.state;
  // Do validation only if DatePicker's popup is dismissed
  if (isDatePickerShown) {
    return;
  }

  // Check when DatePicker is a required field but has NO input value
  if (isRequired && !formattedDate) {
    this.setState({
      errorMessage,
    });
  }
};


export const Picker = styled(DatePicker)`
  ${(props) => props.disabled && '.ms-Icon { display: none; }'}
`;

export class FieldDate extends React.Component {
  handleChange = (date) => {
    const { input } = this.props;
    input.onChange(date);
  }

  render() {
    const {
      meta,
      input,
      required,
      ...props
    } = this.props;

    const { touched, error } = meta;
    const { value } = input;
    const errorMessage = touched && error ? error : null;

    const fieldProps = {
      ...props,
      value,
      errorMessage,
      isRequired: required,
      highlightCurrentMonth: true,
      onSelectDate: this.handleChange,
    };

    return (
      <Picker {...fieldProps} />
    );
  }
}

FieldDate.propTypes = {
  meta: PropTypes.object.isRequired,
  input: PropTypes.object.isRequired,
  required: PropTypes.bool,
};

export default Field(FieldDate, isEmptyDate);
