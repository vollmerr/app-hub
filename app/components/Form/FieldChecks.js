import React from 'react';
import PropTypes from 'prop-types';
import { Checkbox } from 'office-ui-fabric-react/lib/Checkbox';
import styled from 'styled-components';
import { Label } from 'office-ui-fabric-react/lib/Label';

import { metaProp, inputProp, optionsProp } from 'utils/propTypes';
import { isEmptyChecks } from 'utils/validate';
import theme from 'utils/theme';

import Field from './Field';
import FieldError from './FieldError';


export const Check = styled(Checkbox)`
  margin: 5px 0;

  .ms-Checkbox-checkbox {
    ${(props) => props.errorMessage && `border-color: ${theme.redDark};`}
  }
`;


export class FieldChecks extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      options: this.props.options,
    };
  }

  componentWillReceiveProps(nextProps) {
    const { input, options } = nextProps;
    // when redux form resets, reset the options to not checked
    if (!input.value) {
      const newOptions = options.map((option) => ({ ...option, checked: false }));
      this.setState({ options: newOptions });
    }
  }

  handleChange = (event, isChecked, key) => {
    const { input } = this.props;
    const { options } = this.state;
    const newOptions = [...options];
    // set the selected checkbox to checked (or not checked if false)
    newOptions.find((option) => option.key === key).checked = isChecked;
    // get just keys that are checked for redux store
    const checked = newOptions
      .filter((option) => option.checked)
      .map((option) => option.key);
    // update in redux store and state
    input.onChange(checked);
    this.setState({ options: newOptions });
  }

  render() {
    const {
      meta,
      input,
      label,
      required,
      ...props
    } = this.props;

    const { options } = this.state;

    const { touched, error } = meta;
    const errorMessage = touched && error ? error : '';

    const labelProps = {
      label,
      required,
    };

    const checkBoxProps = {
      ...props,
      errorMessage,
    };

    return (
      <div>
        {label && <Label {...labelProps}>{label}</Label>}
        {
          options.map((option) => (
            <Check
              {...option}
              {...checkBoxProps}
              label={option.text}
              onChange={(event, checked) => this.handleChange(event, checked, option.key)}
            />
          ))
        }
        {errorMessage && <FieldError>{errorMessage}</FieldError>}
      </div>
    );
  }
}


const { bool, string } = PropTypes;

FieldChecks.propTypes = {
  meta: metaProp.isRequired,
  input: inputProp.isRequired,
  label: string,
  required: bool,
  options: optionsProp.isRequired,
};

export default Field(FieldChecks, isEmptyChecks);
