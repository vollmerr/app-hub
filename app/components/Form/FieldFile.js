import React from 'react';
import PropTypes from 'prop-types';
import Dropzone from 'react-dropzone';
import styled from 'styled-components';
import { DefaultButton, IconButton } from 'office-ui-fabric-react/lib/Button';
import { Label } from 'office-ui-fabric-react/lib/Label';

import { metaProp, inputProp } from 'utils/propTypes';
import { isEmptyFiles } from 'utils/validate';
import theme from 'utils/theme';

import Field from './Field';
import FieldError from './FieldError';


export const FilePicker = styled.div`
  display: flex;
  justify-content: space-between;
  border: 1px solid;
  border-color: ${(props) => (
    (props.disabled && `${theme.neutralLight}`) ||
    (props['data-error'] && `${theme.redDark}`) ||
    `${theme.neutralTertiary}`
  )};

  ${(props) => !props.disabled &&
    `&:hover {
      border-color: ${theme.neutralPrimary};
    }`
  }
`;


export const DropZone = styled(Dropzone)`
  flex: 1;
  display: flex;
  align-items: center;
  margin-right: auto;
  height: 32px;
  cursor: pointer;
`;


export const FileName = styled.div`
  padding: 0 15px;
  color: ${(props) => (
    props.disabled && `${theme.neutralTertiary}`
  )};
`;


export class FieldFile extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      files: [],
      attachError: null,
    };
  }

  componentWillReceiveProps(nextProps) {
    const { input } = nextProps;
    // when redux form resets, reset the options to not checked
    if (!input.value || !input.value.length) {
      this.setState({ files: [], attachError: null });
    }
  }

  onDrop = (acceptedFiles, rejectedFiles) => {
    const { input } = this.props;

    let files = [];
    let attachError = null;
    if (acceptedFiles.length) {
      files = acceptedFiles;
    } else {
      files = rejectedFiles;
      attachError = 'Failed to attch file.';
    }

    this.setState({ files, attachError });
    input.onChange(files);
  }

  onClear = () => {
    const { input } = this.props;
    const files = [];
    this.setState({ files, attachError: null });
    input.onChange(files);
  }

  // handleFileDownload = () => {
  //   if (this.state.files.length) {
  //     // make into blob and save
  //     const a = document.createElement('a');
  //     const data = [this.state.files[0]];
  //     const name = this.state.files[0].name;
  //     const blob = new Blob(data, { type: 'octet/stream' });

  //     document.body.appendChild(a);
  //     // IE...
  //     if (window.navigator.msSaveOrOpenBlob) {
  //       window.navigator.msSaveOrOpenBlob(blob, name);
  //     } else {
  //       const url = window.URL.createObjectURL(blob);

  //       a.href = url;
  //       a.download = name;
  //       a.click();
  //       window.URL.revokeObjectURL(url);
  //     }
  //   }
  // }

  render() {
    const {
      meta,
      input,
      label,
      disabled,
      required,
      placeholder,
      ...props
    } = this.props;

    const { attachError, files } = this.state;
    const hasFiles = !!files.length && !disabled;
    const { touched, error } = meta;
    const name = this.state.files[0] ? this.state.files[0].name : placeholder || 'Select a file to upload';

    let errorMessage;
    if (attachError) {
      errorMessage = attachError;
    } else if (touched && error) {
      errorMessage = error;
    }

    const labelProps = {
      label,
      required,
    };

    const fieldProps = {
      disabled,
      'data-error': errorMessage,
    };

    /* istanbul ignore next: TODO: how to test refs... */
    const zoneProps = {
      ...props,
      type: 'file',
      disabled,
      'data-error': errorMessage,
      onDrop: this.onDrop,
      innerRef: (node) => { this.dropZoneRef = node; },
    };

    const nameProps = {
      disabled: disabled || !this.state.files.length,
      children: name,
    };

    const clearProps = {
      iconProps: {
        iconName: 'Clear',
      },
      onClick: this.onClear,
    };

    /* istanbul ignore next: TODO: how to test refs... */
    const buttonProps = {
      primary: true,
      disabled,
      type: 'button',
      text: 'Attach File',
      onClick: () => { this.dropZoneRef.open(); },
    };

    return (
      <div>
        {label && <Label {...labelProps}>{label}</Label>}


        <FilePicker {...fieldProps}>
          {hasFiles && <IconButton {...clearProps} />}
          <DropZone {...zoneProps}>
            <FileName {...nameProps} />
          </DropZone>
          <DefaultButton {...buttonProps} />
        </FilePicker>

        {errorMessage && <FieldError>{errorMessage}</FieldError>}

        {/* <button onClick={this.handleFileDownload}>Download</button> */}
      </div>
    );
  }
}


const { bool, string, any } = PropTypes;

FieldFile.propTypes = {
  meta: metaProp.isRequired,
  input: inputProp(any).isRequired,
  label: string,
  disabled: bool,
  required: bool,
  placeholder: string,
};

export default Field(FieldFile, isEmptyFiles);
