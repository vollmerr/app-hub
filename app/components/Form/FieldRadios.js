import React from 'react';
import PropTypes from 'prop-types';
import { ChoiceGroup } from 'office-ui-fabric-react/lib/ChoiceGroup';
import styled from 'styled-components';

import { isEmptyText } from 'utils/validate';
import theme from 'utils/theme';

import Field from './Field';
import FieldError from './FieldError';

export const Radios = styled(ChoiceGroup)`
  .ms-ChoiceField {
    margin: 0;
  }

  .ms-ChoiceField-field::before {
    ${(props) => props.errorMessage && `border-color: ${theme.redDark};`}
  }
`;


export class FieldRadios extends React.Component {
  handleChange = (event, option) => {
    const { input } = this.props;
    input.onChange(option.key);
  }

  render() {
    const {
      meta,
      input,
      ...props
    } = this.props;

    const { touched, error } = meta;
    const { value } = input;
    const errorMessage = touched && error ? error : '';

    const fieldProps = {
      ...props,
      errorMessage,
      selectedKey: value,
      onChange: this.handleChange,
    };

    return (
      <div>
        <Radios {...fieldProps} />
        {errorMessage && <FieldError>{errorMessage}</FieldError>}
      </div>
    );
  }
}

FieldRadios.propTypes = {
  meta: PropTypes.object.isRequired,
  input: PropTypes.object.isRequired,
};

export default Field(FieldRadios, isEmptyText);
