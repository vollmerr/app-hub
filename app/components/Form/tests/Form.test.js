import React from 'react';
import { shallow } from 'enzyme';
import { testStyledComponent } from 'utils/testUtils';
import 'jest-styled-components';

import Form from '../Form';


testStyledComponent(Form);

const props = {
  vh: 100,
  margin: 15,
};


describe('<Form />', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Form {...props} />);
  });

  it('should render correctly with vh and margin passed', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with nothing passed', () => {
    wrapper.setProps({ vh: undefined, margin: undefined });
    expect(wrapper).toMatchSnapshot();
  });
});
