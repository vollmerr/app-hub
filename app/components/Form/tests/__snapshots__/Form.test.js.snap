// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> should render correctly with nothing passed 1`] = `
.c0 {
  min-height: calc(100vh - 50px - 30px - 0px);
  padding: 15px;
  margin: 15px 0;
  background: #ffffff;
}

<form
  className="c0"
/>
`;

exports[`<Form /> should render correctly with vh and margin passed 1`] = `
.c0 {
  min-height: calc(100vh - 50px - 30px - 15px);
  padding: 15px;
  margin: 15px 0;
  background: #ffffff;
}

<form
  className="c0"
/>
`;

exports[`<Form__Form /> should render correctly 1`] = `
.c0 {
  min-height: calc(100vh - 50px - 30px - 0px);
  padding: 15px;
  margin: 15px 0;
  background: #ffffff;
}

<form
  className="c0"
/>
`;
